### PR TITLE
Recover when user has progress on a deleted script

### PR DESCRIPTION
**Pull request by islemaster**
The levelbuilder homepage has been failing to render for some users because their user accounts have progress on a deleted script - that is, they still have a `user_script` entry but the associated `script` is gone.  This PR updates the `User::in_progress_and_deleted_scripts` method (which is only used when rendering the dashboard homepage for logged-in users) to skip any `user_script` entries for which an associated `script` cannot be found. ([Related discussion in Slack](https://codedotorg.slack.com/archives/dashboard-dev/p1480986991000002))

I am slightly concerned about generating additional queries here but since this helper method is only used once on the dashboard homepage and scripts are cached I think this is fine.

Repro'd by manually adding a bad `user_scripts` row to my local dashboard db.

Before:
screenshot from 2016-12-06 15-25-49

After:
screenshot from 2016-12-06 15-26-16

### CODE REVIEW

**Comment by ashercodeorg** [comment]
Unit test?  :)

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -731,7 +731,15 @@ def advertised_scripts
   def in_progress_and_completed_scripts
     backfill_user_scripts if needs_to_backfill_user_scripts?
 
-    user_scripts.compact
+    user_scripts.compact.reject do |user_script|
+      begin
+        user_script.script.nil?
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb:736]: Can this DB query be avoided by doing a lookup in the script cache for `user_script.script_id`?

Comment (islemaster) [reply dashboard/app/models/user.rb:736]: `UserScript.script` does exactly that under the hood ([link](https://github.com/code-dot-org/code-dot-org/blob/571b0750e50d14148af5beda90e6d621946870b5/dashboard/app/models/user_script.rb#L26-L26)); is it necessary to do so here as well?

```ruby
# user_script.rb:25
  def script
    Script.get_from_cache(script_id)
  end
```

Comment (aoby) [reply dashboard/app/models/user.rb:736]: And then if it's not in the cache, [Script.get_from_cache](https://github.com/code-dot-org/code-dot-org/blob/2bd331c55674584f572158bfad2c48857ecdadf2/dashboard/app/models/script.rb#L283-L287) does a DB lookup.

```ruby
def self.get_from_cache(id)
  return get_without_cache(id) unless should_cache?

  script_cache[id.to_s] || get_without_cache(id)
end
```

so we could directly check `Script.script_cache`, though that might be overkill.

Comment (islemaster) [reply dashboard/app/models/user.rb:736]: In ~a production~ each environment do we expect all scripts to be in the `script_cache` when the server starts?  If so, I think `get_from_cache` is sufficient (you'll only hit the DB in the hopefully rare bad case) but either will work; if not, looking in `script_cache` directly will sometimes cause you to skip valid scripts.

File: dashboard/test/models/user_test.rb [hunk 2]
```diff
@@ -574,6 +574,28 @@ class UserTest < ActiveSupport::TestCase
     assert student.reload.encrypted_password != old_password
   end
 
+  test 'user in_progress_and_completed_scripts does not include deleted scripts' do
+    user = create :user
+    real_script = Script.starwars_script
+    fake_script = create :script
+
+    user_script_1 = create :user_script, user: user, script: real_script
+    user_script_2 = create :user_script, user: user, script: fake_script
```
Comment (islemaster) [inline dashboard/test/models/user_test.rb:583]: I assume I don't need to do manual cleanup of records created via FactoryGirl?  It doesn't look like we do elsewhere.

File: dashboard/app/models/user.rb (hunk 1 above)
Comment (ashercodeorg) [reply dashboard/app/models/user.rb:736]: My fault, I assumed a DB query was being generated based on the PR description (that mentions additional DB call). Given that `UserScript.script` uses the cache, LGTM!

File: dashboard/test/models/user_test.rb (hunk 2 above)
Comment (aoby) [reply dashboard/test/models/user_test.rb:583]: Correct

File: dashboard/test/models/user_test.rb [hunk 3]
```diff
@@ -574,6 +574,28 @@ class UserTest < ActiveSupport::TestCase
...
+
+    fake_script.destroy!
+
+    # Preconditions for test: The script is gone, but the associated UserScript still exists.
+    # If we start failing this setup assertion (that is, we do automated cleanup
+    # when deleting a script) then we can probably delete this test.
+    refute Script.exists? fake_script.id
```
Comment (aoby) [inline dashboard/test/models/user_test.rb]: You could also include a message (3rd arg) explaining the failure.

**Pull request merged**